### PR TITLE
Handle datetime selection in cuDF interface

### PR DESCRIPTION
**Pull request by philippjfr**
cuDF does not properly support datetime comparisons so for now we have to cast the datetimes to integers to safely do the comparisons.

### CODE REVIEW

**Comment by hoxbro** [comment]
I'm not sure I follow why it does not support correct selection:

image

``` python
import cudf
from datetime import datetime

cdf = cudf.DataFrame(cudf.date_range("2020-01-01", "2021-01-01")) 

cdf < datetime(2020, 2, 1)
```

Though, I can see this fails: `cdf.values` with `NotImplementedError: DateTime Arrays is not yet implemented in cudf`

**Comment by philippjfr** [comment]
Hmm, internally we do perform the comparisons on the underlying arrays I think. Here we may have to change that.

**Comment by philippjfr** [comment]
Actually, sorry, the traceback is this:

```
  File "/home/philippjfr/development/holoviews/holoviews/core/data/cudf.py", line 261, in select
    selection_mask = cls.select_mask(dataset, selection)
  File "/home/philippjfr/development/holoviews/holoviews/core/data/cudf.py", line 212, in select_mask
    new_masks.append(start <= arr)
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/cudf/core/indexed_frame.py", line 3500, in __array_ufunc__
    ret = super().__array_ufunc__(ufunc, method, *inputs, **kwargs)
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/cudf/core/frame.py", line 1772, in __array_ufunc__
    return _array_ufunc(self, ufunc, method, inputs, kwargs)
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/cudf/utils/utils.py", line 93, in _array_ufunc
    return getattr(obj, op)(other)
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/cudf/core/mixins/mixin_factory.py", line 11, in wrapper
    return method(self, *args1, *args2, **kwargs1, **kwargs2)
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/cudf/core/indexed_frame.py", line 3475, in _binaryop
    ColumnAccessor(type(self)._colwise_binop(operands, op)),
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/nvtx/nvtx.py", line 116, in inner
    result = func(*args, **kwargs)
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/cudf/core/frame.py", line 1761, in _colwise_binop
    else getattr(operator, fn)(left_column, right_column)
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/cudf/core/mixins/mixin_factory.py", line 11, in wrapper
    return method(self, *args1, *args2, **kwargs1, **kwargs2)
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/cudf/core/column/datetime.py", line 405, in _binaryop
    other = self._wrap_binop_normalization(other)
  File "/home/philippjfr/miniconda3/envs/cuda_12.3/lib/python3.10/site-packages/cudf/core/column/column.py", line 609, in _wrap_binop_normalization
    other = other.dtype.type(other.item())
ValueError: Converting an integer to a NumPy datetime requires a specified unit
```

**Comment by hoxbro** [comment]
What is your cudf version?

**Comment by philippjfr** [comment]
Also thought it was that, but just tested with 2024.10.1 and still see the first error I posted.

**Comment by hoxbro** [comment]
I'm having a hard time creating a `DatetimeColumn.` Would you happen to have a small example?

**Comment by philippjfr** [comment]
It's ordering dependent!

`cudf.Series < np.datetime64` works.

`np.datetime64 > np.datetime64` does not.

I can just flip the conditions.

File: holoviews/core/data/cudf.py [hunk 1]
```diff
@@ -204,10 +204,20 @@ def select_mask(cls, dataset, selection):
...
+                    ):
+                        arr = arr.astype(int)
+                    if start is not None:
+                        if isinstance(start, util.datetime_types):
+                            start = np.array([start]).astype(int)[0]
+                        new_masks.append(start <= arr)
+                    if end is not None:
+                        if isinstance(end, util.datetime_types):
+                            end = np.array([end]).astype(int)[0]
+                        new_masks.append(arr < end)
```
Comment (philippjfr) [inline holoviews/core/data/cudf.py:220]:
```suggestion
                    if sel.start is not None:
                        new_masks.append(arr >= self.start)
                    if sel.stop is not None:
                        new_masks.append(arr < sel.stop)
```

File: holoviews/core/data/cudf.py [hunk 2]
```diff
@@ -205,7 +205,7 @@ def select_mask(cls, dataset, selection):
                 with warnings.catch_warnings():
                     warnings.filterwarnings('ignore', r'invalid value encountered')
                     if sel.start is not None:
-                        new_masks.append(sel.start <= arr)
+                        new_masks.append(arr >= self.start)
```
Comment (philippjfr) [inline holoviews/core/data/cudf.py:208]:
```suggestion
                        # Comparison has to be in this order due to issues with datetime comparison (see #6407)
                        new_masks.append(arr >= sel.start)
```

File: holoviews/core/data/cudf.py [hunk 3]
```diff
@@ -232,6 +233,7 @@ def select_mask(cls, dataset, selection):
                 mask &= new_mask
         return mask
 
+
```
Comment (hoxbro) [inline holoviews/core/data/cudf.py:236]:
```suggestion
```

**Pull request merged**